### ISSUE TITLE
Support `text_signature` and autogenerated text signature for `#[new]`

### ISSUE THREAD

**Issue by davidhewitt** [PyO3/pyo3#2866]
After #2784 we now support autogenerated `text_signature` for everything except `#[new]`, which can't have any text signature set (it must be put on the `#[pyclass]` instead).

Using some macro and trait magic we should be able to resolve this; I won't do it for 0.18 but perhaps shortly after if nobody else wants to pick this up first (in which case I can point you in the right direction for the implementation).
